asyncpg/connect_utils.py

```diff
@@ -303,6 +314,34 @@ def _parse_connect_dsn_and_args(*, dsn, host, port, user,
         raise ValueError(
             'could not determine the database address to connect to')
 
+    if ssl is None:
+        ssl = os.getenv('PGSSLMODE')
+
+    if isinstance(ssl, str):
+        SSLMODES = {
+          'disable':   0, 'allow':       1,
+          'prefer':    2, 'require':     3,
+          'verify-ca': 4, 'verify-full': 5,
+        }
+        try:
+            sslmode = SSLMODES[ssl]
+        except KeyError:
+            raise ValueError(
+                '`sslmode` parameter must be one of '
+                + ', '.join(SSLMODES.keys()))
+
+        # docs at https://www.postgresql.org/docs/10/static/libpq-connect.html
+        # Not implemented: sslcert & sslkey & sslrootcert & sslcrl params, and
+        # the "retry with/without SSL" behavior for sslmode 'allow' & 'prefer'.
```

Comment:
It doesn't seem hard to actually implement support for the `'prefer'` mode.  An ssl connection should be attempted, but an `N` response here:

https://github.com/MagicStack/asyncpg/blob/ffd134e8168a825cff60e49fa2562cdceb59ee1d/asyncpg/connect_utils.py#L450-L455

should not cause a `ConnectionError`